File: handarm/filters.py
"""Signal smoothing: One Euro filters for position and orientation.

The One Euro filter (Casiez et al., 2012) adapts its cutoff with signal
speed: heavy smoothing when the hand hovers (kills jitter), light smoothing
during fast moves (kills lag). It is the standard choice for pointing and
tracking interfaces.
"""

from __future__ import annotations

import numpy as np

from .transforms import quat_angle_between, quat_normalize, quat_slerp


def _smoothing_factor(dt: float, cutoff: float) -> float:
    r = 2.0 * np.pi * cutoff * dt
    return r / (r + 1.0)


class OneEuroFilter:
    """Vector-valued One Euro filter."""

    def __init__(self, min_cutoff: float = 1.0, beta: float = 0.0, d_cutoff: float = 1.0):
        self.min_cutoff = min_cutoff
        self.beta = beta
        self.d_cutoff = d_cutoff
        self._x = None
        self._dx = None

    def reset(self) -> None:
        self._x = None
        self._dx = None

    def __call__(self, x: np.ndarray, dt: float) -> np.ndarray:
        x = np.asarray(x, dtype=float)
        if self._x is None:
            self._x = x.copy()
            self._dx = np.zeros_like(x)
            return x.copy()
        # A zero/negative dt (coarsened clocks) must not wipe filter memory;
        # hold the previous estimate instead of passing raw jitter through.
        if dt <= 0:
            return self._x.copy()

        # Derivative estimate, low-passed.
        dx = (x - self._x) / dt
        a_d = _smoothing_factor(dt, self.d_cutoff)
        self._dx = a_d * dx + (1 - a_d) * self._dx

        # Speed-adaptive cutoff.
        cutoff = self.min_cutoff + self.beta * float(np.linalg.norm(self._dx))
        a = _smoothing_factor(dt, cutoff)
        self._x = a * x + (1 - a) * self._x
        return self._x.copy()


class QuaternionLowPass:
    """One Euro filter on orientation: slerp toward the target with a cutoff
    that opens up with angular speed. beta = 0 degrades to a plain
    first-order low-pass."""

    def __init__(self, cutoff: float = 3.0, beta: float = 0.0, d_cutoff: float = 1.0):
        self.cutoff = cutoff
        self.beta = beta
        self.d_cutoff = d_cutoff
        self._q = None
        self._rate = 0.0  # low-passed angular speed of the input, rad/s

    def reset(self) -> None:
        self._q = None
        self._rate = 0.0

    def __call__(self, q: np.ndarray, dt: float) -> np.ndarray:
        q = quat_normalize(q)
        if self._q is None:
            self._q = q.copy()
            return q.copy()
        if dt <= 0:
            return self._q.copy()
        a_d = _smoothing_factor(dt, self.d_cutoff)
        self._rate = a_d * (quat_angle_between(self._q, q) / dt) + (1 - a_d) * self._rate
        a = _smoothing_factor(dt, self.cutoff + self.beta * self._rate)
        self._q = quat_slerp(self._q, q, a)
        return self._q.copy()


class ScalarLowPass:
    """Simple exponential moving average with a time-constant cutoff."""

    def __init__(self, cutoff: float = 5.0):
        self.cutoff = cutoff
        self._y = None

    def reset(self) -> None:
        self._y = None

    def __call__(self, y: float, dt: float) -> float:
        if self._y is None:
            self._y = float(y)
            return self._y
        if dt <= 0:
            return self._y
        a = _smoothing_factor(dt, self.cutoff)
        self._y = a * float(y) + (1 - a) * self._y
        return self._y

File: web/src/filters.ts
/**
 * Signal smoothing — TypeScript port of handarm/filters.py.
 *
 * One Euro filter (Casiez et al., 2012): speed-adaptive cutoff — heavy
 * smoothing when the hand hovers, light smoothing during fast moves.
 */

import { type Quat, quatAngleBetween, quatNormalize, quatSlerp } from "./transforms";

function smoothingFactor(dt: number, cutoff: number): number {
  const r = 2 * Math.PI * cutoff * dt;
  return r / (r + 1);
}

export class OneEuroFilter {
  private x: number[] | null = null;
  private dx: number[] | null = null;

  constructor(
    private minCutoff = 1.0,
    private beta = 0.0,
    private dCutoff = 1.0,
  ) {}

  reset(): void {
    this.x = null;
    this.dx = null;
  }

  apply(x: number[], dt: number): number[] {
    if (this.x === null || this.dx === null) {
      this.x = [...x];
      this.dx = x.map(() => 0);
      return [...x];
    }
    // A zero/negative dt (coarsened clocks) must not wipe filter memory;
    // hold the previous estimate instead of passing raw jitter through.
    if (dt <= 0) return [...this.x];
    const aD = smoothingFactor(dt, this.dCutoff);
    let speedSq = 0;
    for (let i = 0; i < x.length; i++) {
      const dxi = (x[i] - this.x[i]) / dt;
      this.dx[i] = aD * dxi + (1 - aD) * this.dx[i];
      speedSq += this.dx[i] * this.dx[i];
    }
    const cutoff = this.minCutoff + this.beta * Math.sqrt(speedSq);
    const a = smoothingFactor(dt, cutoff);
    for (let i = 0; i < x.length; i++) {
      this.x[i] = a * x[i] + (1 - a) * this.x[i];
    }
    return [...this.x];
  }
}

/**
 * One Euro filter on orientation: slerp toward the target with a cutoff
 * that opens up with angular speed. beta = 0 degrades to a plain
 * first-order low-pass.
 */
export class QuaternionLowPass {
  private q: Quat | null = null;
  private rate = 0; // low-passed angular speed of the input, rad/s

  constructor(
    private minCutoff = 3.0,
    private beta = 0.0,
    private dCutoff = 1.0,
  ) {}

  reset(): void {
    this.q = null;
    this.rate = 0;
  }

  apply(q: Quat, dt: number): Quat {
    q = quatNormalize(q);
    if (this.q === null) {
      this.q = [...q];
      return [...q];
    }
    if (dt <= 0) return [...this.q];
    const aD = smoothingFactor(dt, this.dCutoff);
    this.rate = aD * (quatAngleBetween(this.q, q) / dt) + (1 - aD) * this.rate;
    const cutoff = this.minCutoff + this.beta * this.rate;
    this.q = quatSlerp(this.q, q, smoothingFactor(dt, cutoff));
    return [...this.q];
  }
}

/** Simple exponential moving average with a time-constant cutoff. */
export class ScalarLowPass {
  private y: number | null = null;

  constructor(private cutoff = 5.0) {}

  reset(): void {
    this.y = null;
  }

  apply(y: number, dt: number): number {
    if (this.y === null) {
      this.y = y;
      return y;
    }
    if (dt <= 0) return this.y;
    const a = smoothingFactor(dt, this.cutoff);
    this.y = a * y + (1 - a) * this.y;
    return this.y;
  }
}

/**
 * Dead zone that travels with the signal (a backlash / slop filter): the
 * output holds until the input strays more than `slop` from it, then
 * follows at that distance. Unlike a fixed dead zone around a neutral
 * value this suppresses small wobble anywhere in the range, which is
 * what an axis that picks up unintended cross-coupling needs; the cost
 * is `slop` of lag, and twice that to reverse direction.
 */
export class SlopFilter {
  private y: number | null = null;

  constructor(private slop = 0) {}

  reset(): void {
    this.y = null;
  }

  apply(x: number): number {
    if (this.y === null) {
      this.y = x;
      return x;
    }
    if (x > this.y + this.slop) this.y = x - this.slop;
    else if (x < this.y - this.slop) this.y = x + this.slop;
    return this.y;
  }
}
